Cordio: Add tests that validates a cordio port.

pan- opened the pull request:
### Description

The first test focuse on the transport by testing that the reset sequence is
correctly sent to the controller and properly received.

The second test validates that the reset sequence fullfill the right state
of the stack and ensure that bluetooth initialization succeed.

### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ ] New target
    [X] Feature
    [ ] Breaking change

pan- committed:
Cordio: Add tests that validates a cordio port.

The first test focuse on the transport by testing that the reset sequence is
correctly sent to the controller and properly received.

The second test validates that the reset sequence fullfill the right state
of the stack and ensure that bluetooth initialization succeed.

donatieng commented on features/FEATURE_BLE/targets/TARGET_CORDIO/TESTS/cordio_hci/driver/main.cpp:
```diff
@@ -0,0 +1,98 @@
+    if (params->error == BLE_ERROR_NONE) {
+        initialization_state = INITIALIZATION_SUCCESS;
+    } else {
+        initialization_state = INITIALIZATION_SUCCESS;
```
The other one ;)

donatieng commented on features/FEATURE_BLE/targets/TARGET_CORDIO/TESTS/cordio_hci/driver/main.cpp:
```diff
@@ -0,0 +1,98 @@
+using namespace utest::v1;
+using mbed::callback;
+
+#define INITIALIZATION_TIMEOUT (10 * 1000)
```
Is it actually acceptable for the initialisation process to take this long?

pan- replied:
Yes, at first I was thinking of setting it to 30 seconds; if the bitrate is low and the controller require a large service pack then initialization can takes a loooooot of time.

donatieng commented on features/FEATURE_BLE/targets/TARGET_CORDIO/TESTS/cordio_hci/driver/main.cpp:
```diff
@@ -0,0 +1,98 @@
+    TEST_ASSERT_NOT_EQUAL(0, hciCoreCb.availBufs);
+
+    uint8_t invalid_le_states[HCI_LE_STATES_LEN] = { 0 };
+    TEST_ASSERT_NOT_EQUAL(0, memcmp(invalid_le_states, hciCoreCb.leStates, HCI_LE_STATES_LEN));
```
Can you just add a quick comment (as `hciCoreCb` is an internal Cordio struct, so not immediately obvious what you're testing here)

donatieng commented on features/FEATURE_BLE/targets/TARGET_CORDIO/TESTS/cordio_hci/transport/main.cpp:
```diff
@@ -0,0 +1,190 @@
+                }
+                break;
+        }
+
```
You should break from the for loop here if `test_result` is `TEST_RESULT_FAILURE`

paul-szczepanek-arm replied:
doesn't the if do that? it'll break (return) on any result change

pan- replied:
Yes, the if does that whenever a result is present (success or failure).

donatieng replied:
OK - there are a few things that confused me, now I understand what you're doing. Few questions/remarks:

- Could you rename `TEST_RESULT_TIMEOUT_FAILURE` to something more explicit (like 'in progress'); it hasn't timed out till proven otherwise
- At the moment, if you don't know the HCI OpCode's value, you skip to the next byte - which would be parameters' length; what's the rationale behind this? Skipping padding bytes? You could interpret the length field (and subsequent parameters) as OpCodes in the current implementation

donatieng commented on features/FEATURE_BLE/targets/TARGET_CORDIO/TESTS/cordio_hci/transport/main.cpp:
```diff
@@ -0,0 +1,190 @@
+                break;
+        }
+
+        if (test_result != TEST_RESULT_TIMEOUT_FAILURE) {
```
Presumably that should be outside of the for loop?

pan- replied:
Why ? there's no guarantee that data in input is at the event boundary and it may fail early.

donatieng commented on features/FEATURE_BLE/targets/TARGET_CORDIO/TESTS/cordio_hci/transport/main.cpp:
```diff
@@ -0,0 +1,190 @@
+}
+
+Case cases[] = {
+    Case("Test reset command", test_reset_command),
```
Could you add a second test that issues multiple reset commands + `initialize()`/`terminate()`?

donatieng reviewed (changes requested):
Thanks a lot Vincent, a few issues as described

pan- committed:
Cordio BLE: Fix test failure when controller initialization failed.

pan- committed:
Cordio BLE: Document in test what fields are expected to be set after init.

pan- committed:
Cordio BLE: Rewrite reset test strategy

The existing test was expecting that the acknowledgement of the reset command would be the first and only event receive. This assumption is false.

As a consequence, the new code parse all incoming packets and raise a flag in the following circumstances:
* a reset packet has been successfully received.
* RX stream is not synchronized
* The status of the reset command is an error.

Another test has been added that send a serie of reset commands.

pan- commented:
@donatieng @paul-szczepanek-arm I've rewritten the logic in the transport test: all incoming packet are parsed and a flag is raised whenever: 
* A reset packet has been received 
* RX is synced with the controller
* The status of the reset packet is incorrect. 

In the process I've added a second test that sends a serie of reset commands and waits for response events.

donatieng reviewed (approved):
Thanks a lot @pan-, this addresses all of my issues 👍

cmonr commented:
/morph build
